pkg/reconciler/v1alpha1/trigger/trigger.go

```diff
@@ -0,0 +1,687 @@
+		return err
+	}
+	t.Status.MarkKubernetesServiceExists()
+
+	_, err = r.reconcileVirtualService(ctx, t, svc)
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to reconcile the VirtualService", zap.Error(err))
+		return err
+	}
+	t.Status.MarkVirtualServiceExists()
+
+	_, err = r.subscribeToBrokerChannel(ctx, t, c, svc)
+	if err != nil {
+		logging.FromContext(ctx).Error("Unable to Subscribe", zap.Error(err))
+		t.Status.MarkNotSubscribed("notSubscribed", "%v", err)
+		return err
+	}
+	t.Status.MarkSubscribed()
+
+	return nil
+}
+
+func (r *reconciler) AddToTriggers(t *v1alpha1.Trigger) {
+	name := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: t.Namespace,
+			Name:      t.Name,
+		},
+	}
+
+	brokerNamespacedName := types.NamespacedName{
+		Namespace: t.Namespace,
+		Name:      t.Spec.Broker}
+
+	// We will be reconciling an already existing Trigger far more often than adding a new one, so
+	// check with a read lock before using the write lock.
+	r.triggersLock.RLock()
+	_, present := r.triggers[brokerNamespacedName][name]
+	r.triggersLock.RUnlock()
+	if present {
+		// Already present in the map.
+		return
+	}
+
+	r.triggersLock.Lock()
+	triggersInBrokerNamespacedName := r.triggers[brokerNamespacedName]
+	if triggersInBrokerNamespacedName == nil {
+		r.triggers[brokerNamespacedName] = make(map[reconcile.Request]struct{})
+		triggersInBrokerNamespacedName = r.triggers[brokerNamespacedName]
+	}
+	triggersInBrokerNamespacedName[name] = dummyValue
+	r.triggersLock.Unlock()
+}
+
+func (r *reconciler) removeFromTriggers(t *v1alpha1.Trigger) {
+	name := reconcile.Request{
+		NamespacedName: types.NamespacedName{
+			Namespace: t.Namespace,
+			Name:      t.Name,
+		},
+	}
+
+	brokerNamespacedName := types.NamespacedName{
+		Namespace: t.Namespace,
+		Name:      t.Spec.Broker}
+
+	r.triggersLock.Lock()
+	triggersInBrokerNamespacedName := r.triggers[brokerNamespacedName]
+	if triggersInBrokerNamespacedName != nil {
+		delete(triggersInBrokerNamespacedName, name)
+	}
+	r.triggersLock.Unlock()
+}
+
+// updateStatus may in fact update the trigger's finalizers in addition to the status.
+func (r *reconciler) updateStatus(trigger *v1alpha1.Trigger) (*v1alpha1.Trigger, error) {
+	ctx := context.TODO()
+	objectKey := client.ObjectKey{Namespace: trigger.Namespace, Name: trigger.Name}
+	latestTrigger := &v1alpha1.Trigger{}
+
+	if err := r.client.Get(ctx, objectKey, latestTrigger); err != nil {
+		return nil, err
+	}
+
+	triggerChanged := false
+
+	if !equality.Semantic.DeepEqual(latestTrigger.Finalizers, trigger.Finalizers) {
+		latestTrigger.SetFinalizers(trigger.ObjectMeta.Finalizers)
+		if err := r.client.Update(ctx, latestTrigger); err != nil {
+			return nil, err
+		}
+		triggerChanged = true
+	}
+
+	if equality.Semantic.DeepEqual(latestTrigger.Status, trigger.Status) {
+		return latestTrigger, nil
+	}
+
+	if triggerChanged {
+		// Refetch
```

Comment:
I'm not sure why this refetch is necessary. Maybe because of the finalizer update above? Anyway I think this update logic is incorrect.

IIUC, the flow here is:

1. Get latest trigger (triggerA)
1. Make updates to triggerA
1. Get latest trigger (triggerB)
1. if triggerB finalizers != triggerA finalizers
  Update triggerB with triggerA's finalizers
1. if triggerA status != triggerB status
  Get latest trigger (triggerC)
1. update triggerC's status using triggerA's status

This will clobber 3rd-party updates to the finalizers made between 1 and 3, and 3rd-party updates to the status between 1 and 6.

I think it's safer to update only once per loop, using the resourceVersion that was originally retrieved from the cache in step 1 (triggerA).

1. Get latest trigger (triggerD)
1. Copy triggerD to triggerE
1. Make updates to triggerE
1. if triggerD finalizers != triggerE finalizers
    update triggerE finalizers
    return
1. else if triggerD != triggerE
    update triggerE status _(note this doesn't happen if the finalizers have been updated)_

Now if there was a change made between 1 and 4 or 5, the update in 4 or 5 will fail because of the resourceVersion conflict. The reconcile attempt will return an error and be retried again, possibly with backoff. The [eventing-sources SDK mostly does this](https://github.com/knative/eventing-sources/blob/0c1c5457cda74a6bdeedbce942ce878fb55e4082/pkg/controller/sdk/reconciler.go), but it doesn't do an early return when a finalizer is updated. The early return when a finalizer is updated eliminates the need to do a double update and refetch. It's fine to wait for another reconcile loop, because the finalizer update will come through the watch soon after.

It might seem like this is only a theoretical concern because reconcile doesn't take very long and the window for conflicting updates is very small, but in fact informer cache latency can be several seconds so the window is wider than just the reconcile time.

Reply:
I created #890 to track this and avoid blocking this PR.

Reply:
We'll leave this as-is for this PR and fix it via #890.